sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -131,181 +164,160 @@ public boolean isEndRequired() {
+    spanBuffer.clear();
+    if (queue.drainTo(spanBuffer) == 0) {
+      scheduleNextExport();
+      return;
     }
 
-    @Override
-    public void run() {
-      while (!Thread.currentThread().isInterrupted()) {
-        // Copy all the batched spans in a separate list to release the monitor lock asap to
-        // avoid blocking the producer thread.
-        ArrayList<ReadableSpan> spansCopy;
-        synchronized (monitor) {
-          // If still maxExportBatchSize elements in the queue better to execute an extra
-          if (spansList.size() < maxExportBatchSize) {
-            do {
-              // In the case of a spurious wakeup we export only if we have at least one span in
-              // the batch. It is acceptable because batching is a best effort mechanism here.
-              try {
-                monitor.wait(scheduleDelayMillis);
-              } catch (InterruptedException ie) {
-                // Preserve the interruption status as per guidance and stop doing any work.
-                Thread.currentThread().interrupt();
-                return;
-              }
-            } while (spansList.isEmpty());
+    final CompletableResultCode result = new CompletableResultCode();
+    executor.submit(
+        new Runnable() {
+          @Override
+          public void run() {
+            exportNextBatch(spanBuffer, 0, result);
           }
-          spansCopy = new ArrayList<>(spansList);
-          spansList.clear();
-        }
-        // Execute the batch export outside the synchronized to not block all producers.
-        exportBatches(spansCopy);
-      }
-    }
-
-    private void shutdown() {
-      forceFlush();
-      timer.cancel();
-      spanExporter.shutdown();
-    }
+        });
+
+    result.whenComplete(
+        new Runnable() {
+          @Override
+          public void run() {
+            if (queue.size() >= maxExportBatchSize) {
+              executor.submit(
+                  new Runnable() {
+                    @Override
+                    public void run() {
+                      exportSpans();
+                    }
+                  });
+            } else {
+              scheduleNextExport();
+            }
+          }
+        });
+  }
 
-    private void forceFlush() {
-      ArrayList<ReadableSpan> spansCopy;
-      synchronized (monitor) {
-        spansCopy = new ArrayList<>(spansList);
-        spansList.clear();
-      }
-      // Execute the batch export outside the synchronized to not block all producers.
-      exportBatches(spansCopy);
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void exportNextBatch(
+      final List<ReadableSpan> spans, final int offset, final CompletableResultCode result) {
+    int num = Math.min(maxExportBatchSize, spans.size() - offset);
+    final int limit = offset + num;
+    List<SpanData> forExport = new ArrayList<>(num);
+    for (int i = offset; i < limit; i++) {
+      forExport.add(spans.get(i).toSpanData());
     }
```

Comment:
I'm thinking that a native array copy would be more efficient here... and then form an array list around that for calling the exporter. WDYT?

Reply:
TBH I've never compared performance of preallocated `ArrayList` with `.add` vs allocating an array and using `Arrays.asList`. I'll need to write a JMH to confirm the difference, since `i` doesn't start at `0` there is some extra cognitive load so I'd like to try that after confirming the perf difference.